### PR TITLE
Make ISIC multimodal again

### PR DESCRIPTION
**Pull request by SagiPolaczek**
Make ISIC multimodal, when the additional modality is:
* Sex - patient's sex
* Age - patient's age
* Anatom site - the location of the lesion on the patient's body

### CODE REVIEW

File: examples/fuse_examples/imaging/classification/isic/isic_runner.py [hunk 1]
```diff
@@ -93,12 +108,12 @@
 TRAIN_COMMON_PARAMS["data.num_folds"] = 5
 TRAIN_COMMON_PARAMS["data.train_folds"] = [0, 1, 2]
 TRAIN_COMMON_PARAMS["data.validation_folds"] = [3]
-TRAIN_COMMON_PARAMS["data.samples_ids"] = FULL_GOLDEN_MEMBERS  # Change to None to use all members
+TRAIN_COMMON_PARAMS["data.samples_ids"] = {"all": None, "golden": FULL_GOLDEN_MEMBERS}["golden"]
```
Comment (mosheraboh) [inline examples/fuse_examples/imaging/classification/isic/isic_runner.py:111]: Why is the default set to golden?

File: examples/fuse_examples/imaging/classification/isic/isic_runner.py [hunk 2]
```diff
@@ -93,12 +108,12 @@
...
 
 # ===============
 # PL Trainer
 # ===============
-TRAIN_COMMON_PARAMS["trainer.num_epochs"] = 20
+TRAIN_COMMON_PARAMS["trainer.num_epochs"] = 10
```
Comment (mosheraboh) [inline examples/fuse_examples/imaging/classification/isic/isic_runner.py:116]: 10 epochs already achieve the best results?

File: examples/fuse_examples/imaging/classification/isic/isic_runner.py [hunk 3]
```diff
@@ -172,6 +200,7 @@ def run_train(paths: dict, train_common_params: dict) -> None:
         output_split_filename=paths["data_split_filename"],
         keys_to_balance=["data.label"],
         nfolds=train_common_params["data.num_folds"],
+        reset_split=True,
```
Comment (mosheraboh) [inline examples/fuse_examples/imaging/classification/isic/isic_runner.py:203]: I wouldn't set it to True by default

File: examples/fuse_examples/tests/test_classification_cmmd.py [hunk 4]
```diff
@@ -54,19 +54,19 @@ def run_cmmd(root: str) -> None:
                 "target": "classification",
                 "reset_cache": False,
                 "num_workers": 10,
```
Comment (mosheraboh) [inline examples/fuse_examples/tests/test_classification_cmmd.py:56]: Why did you change cmmd?

File: fuse/data/ops/ops_aug_tabular.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+import numpy as np
+from typing import Optional, Any, Sequence
+
+from fuse.data.ops.op_base import OpBase
+
+from fuse.utils import NDict
+
+
+class OpAugOneHot(OpBase):
+    """
```
Comment (mosheraboh) [inline fuse/data/ops/ops_aug_tabular.py:11]: Cool, just update the documentation - cause now you don't get prob.
Instead, show examples with OpRandApply.

File: fuse/data/ops/ops_aug_tabular.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    def __call__(
+        self,
+        sample_dict: NDict,
+        key: str,
+        idx: Optional[int] = None,
+        freeze_indices: Sequence[int] = [],
+        mode: str = "default",
+    ) -> NDict:
+        """
+        :param key: key for the one-hot vector
```
Comment (mosheraboh) [inline fuse/data/ops/ops_aug_tabular.py:40]: Specify here that you expect numpy array

File: fuse/data/ops/ops_aug_tabular.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+            # idx +- 1 with probability of 0.5
+            idx = (idx + 1) if random.random() < 0.5 else (idx - 1)
+
+        # make sure idx in range
+        idx = max(idx, 0)
+        idx = min(idx, len(one_hot) - 1)
+
+        # set one-hot
+        res_one_hot[idx] = 1
+        sample_dict[key] = one_hot
```
Comment (mosheraboh) [inline fuse/data/ops/ops_aug_tabular.py:75]: sample_dict[key] = res_one_hot?

File: fuse/dl/lightning/pl_module.py [hunk 8]
```diff
@@ -134,7 +134,7 @@ def training_epoch_end(self, step_outputs) -> None:
 
     def validation_epoch_end(self, step_outputs) -> None:
         # for the logs to be at each epoch, not each step
-        self.log("step", self.current_epoch, on_epoch=True)
+        self.log("step", float(self.current_epoch), on_epoch=True)
```
Comment (mosheraboh) [inline fuse/dl/lightning/pl_module.py:137]: Why?

**Review by mosheraboh — CHANGES_REQUESTED** [review]
Thanks, looks good.
Few comments inline.

File: fuse/dl/lightning/pl_module.py (hunk 8 above)
Comment (SagiPolaczek) [reply fuse/dl/lightning/pl_module.py:137]: Solves the following warning:

```
<packages_path>/pytorch_lightning/trainer/connectors/logger_connector/result.py:229: UserWarning: You called `self.log('step', ...)` in your `validation_epoch_end` but the value needs to be floating point. Converting it to torch.float32.
```

Now I double checked that, seems I missed the same fix in the `training_epoch_end` function.
Sure we can ignore it :-)

File: examples/fuse_examples/imaging/classification/isic/isic_runner.py (hunk 2 above)
Comment (SagiPolaczek) [reply examples/fuse_examples/imaging/classification/isic/isic_runner.py:116]: No. I set the runner for a quick run session (also the reason for `golden` as default.

I'm just about to deploy several final experiments and I'll set it as the best one.

File: fuse/data/ops/ops_aug_tabular.py (hunk 7 above)
Comment (SagiPolaczek) [reply fuse/data/ops/ops_aug_tabular.py:75]: thanks nice catch 😄  prob missed that in one of the changes..

File: examples/fuse_examples/tests/test_classification_cmmd.py (hunk 4 above)
Comment (SagiPolaczek) [reply examples/fuse_examples/tests/test_classification_cmmd.py:56]: Forgot to mention but I played with it to speed-up the CI test (less data used and less epochs).
It's just a WA tho.. What do you think?

**Review by SagiPolaczek — COMMENTED** [review]
@mosheraboh 

I'm executing final runs and I'll edit again once finishing. I'll notify when its done.

File: fuse/data/ops/ops_aug_tabular.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+        return sample_dict
+
+
+class OpAugReplaceWithProb(OpBase):
```
Comment (SagiPolaczek) [inline fuse/data/ops/ops_aug_tabular.py:80]: deleting it

File: fuse/dl/lightning/pl_module.py (hunk 8 above)
Comment (mosheraboh) [reply fuse/dl/lightning/pl_module.py:137]: Great, thanks

File: examples/fuse_examples/tests/test_classification_cmmd.py (hunk 4 above)
Comment (mosheraboh) [reply examples/fuse_examples/tests/test_classification_cmmd.py:56]: Seems good.Lets keep it,
I think that the ccc command requesting cpu cores and specific GPU will fix it.

File: fuse/utils/gpu.py [hunk 10]
```diff
@@ -168,9 +168,9 @@ def get_available_gpu_ids_from_nvidia_smi_output(raw_output: str) -> List[int]:
         if "Off" in line:
             current_gpu = int(split_output[line_idx][4])
             continue
-        if current_gpu is not -1:
+        if current_gpu != -1:
```
Comment (SagiPolaczek) [inline fuse/utils/gpu.py:171]: Avoid warnings

File: fuse/utils/gpu.py [hunk 11]
```diff
@@ -168,9 +168,9 @@ def get_available_gpu_ids_from_nvidia_smi_output(raw_output: str) -> List[int]:
...
             loc = line.find("MiB /")
-            if loc is not -1:
+            if loc != -1:
```
Comment (SagiPolaczek) [inline fuse/utils/gpu.py:173]: same

File: fuseimg/datasets/tests/test_datasets.py [hunk 12]
```diff
@@ -92,7 +92,7 @@ def test_isic(self):
 
         create_dir(self.isic_cache_dir)
         dataset = ISIC.dataset(
-            self.isic_data_dir, self.isic_cache_dir, reset_cache=True, samples_ids=TEN_GOLDEN_MEMBERS
+            self.isic_data_dir, self.isic_cache_dir, train=True, reset_cache=True, samples_ids=TEN_GOLDEN_MEMBERS
```
Comment (SagiPolaczek) [inline fuseimg/datasets/tests/test_datasets.py:95]: to apply all the pipeline

File: fuse/data/README.md [hunk 13]
```diff
@@ -309,13 +324,15 @@ The following operators are useful when implementing a common pipeline:
 * OpToTensor - convert many different types to PyTorch tensor
 * OpOneHotToNumber - convert one-hot encoding vectors into numbers
 
+[**Tabular augmentations**](ops/ops_aug_tabular.py)
+
+* OpAugOneHot - Apply an augmentation for an one-hot encoding vector
```
Comment (SagiPolaczek) [inline fuse/data/README.md:329]: Added the Op here. 

All the other changes are result of some linter.. not on purpose but looks good.

File: run_all_unit_tests.sh [hunk 14]
```diff
@@ -30,7 +30,8 @@ create_env() {
         requirements+=$(cat examples/requirements.txt)
     fi
 
-    PYTHON_VER=3.7
+    # Python version
+    PYTHON_VER=3.9
```
Comment (mosheraboh) [inline run_all_unit_tests.sh:34]: 3.8?

**Review by mosheraboh — APPROVED** [review]
Great!
Just switch to python 3.8 and merge.
